DigitalLearningSolutions.Data/DataServices/ProgressDataService.cs

```diff
@@ -227,5 +235,98 @@ public void UnlockProgress(int progressId)
                 new { progressId }
             );
         }
+
+        public Progress? GetProgressByProgressId(int progressId)
+        {
+            return connection.Query<Progress>(
+                @"SELECT ProgressId,
+                        CandidateID,
+                        CustomisationID,
+                        Completed,
+                        RemovedDate,
+                        SystemRefreshed,
+                        SupervisorAdminID,
+                        CompleteByDate,
+                        EnrollmentMethodID,
+                        EnrolledByAdminID,
+                        SubmittedTime,
+                        CustomisationVersion,
+                        DiagnosticScore
+                    FROM Progress
+                    WHERE ProgressID = @progressId",
+                new { progressId }
+            ).SingleOrDefault();
+        }
+
+        public IEnumerable<DetailedSectionProgress> GetSectionProgressDataForProgressEntry(int progressId)
+        {
+            return connection.Query<DetailedSectionProgress>(
+                @"SELECT
+                        s.SectionID,
+                        s.SectionName,
+                        (SUM(asp1.TutStat) * 100) / (COUNT(t.TutorialID) * 2) AS Completion,
+                        SUM(asp1.TutTime) AS TotalTime,
+                        s.AverageSectionMins AS AverageTime,
+                        CASE WHEN (s.PLAssessPath IS NOT NULL AND cu.IsAssessed = 1) THEN 1 ELSE 0 END AS PostLearningAssessment,
+                        COALESCE (MAX(ISNULL(aa.Score, 0)), 0) AS Outcome,
+                        (SELECT COUNT(AssessAttemptID) AS PLAttempts
+                            FROM AssessAttempts AS aa
+                            WHERE (ProgressID = @ProgressID) AND (SectionNumber = s.SectionNumber)) AS Attempts,
+                        MAX(ISNULL(CAST(ct.Status AS INT), 0)) AS Passed
+                    FROM
+                        aspProgress AS asp1
+                        INNER JOIN Progress AS p ON asp1.ProgressID = p.ProgressID
+                        INNER JOIN Sections AS s
+                        INNER JOIN Tutorials AS t ON s.SectionID = t.SectionID
+                        INNER JOIN CustomisationTutorials AS ct ON t.TutorialID = ct.TutorialID ON asp1.TutorialID = t.TutorialID
+                        INNER JOIN Customisations AS cu ON p.CustomisationID = cu.CustomisationID
+                        LEFT OUTER JOIN AssessAttempts AS aa ON p.ProgressID = aa.ProgressID AND s.SectionNumber = aa.SectionNumber

```

Comment:
The order of tables/joins looks weird to me and took me a second to discern. Think I am used to seeing a linear set of connections. Is this ordering intentional, and is it worth changing it around to fit in with the other data service query orders?

Reply:
Like many of the oddities here it's inherited directly from the pre-existing SQL procedure in the legacy system, which I erred on the side of not meddling with. I'll see if I can disambiguate.

Reply:
Having looked around I think this ordering is intentional, to control the JOIN order: https://www.sqlservercentral.com/forums/topic/multiple-on-clauses-in-one-join#post-1723035